ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -55,60 +56,60 @@ where
         Self { u, v }
     }
 
-    pub fn compute_proof<λ: ArrayLength>(
-        &self,
-        r: F,
-    ) -> (ZeroKnowledgeProof<F, TwoNMinusOne<λ>>, ProofGenerator<F>)
+    pub fn compute_proof<λ: ArrayLength, I, J>(
+        u: I,
+        v: J,
+    ) -> ZeroKnowledgeProof<F, TwoNMinusOne<λ>>
     where
         λ: ArrayLength + Add + Sub<U1>,
         <λ as Add>::Output: Sub<U1>,
         <<λ as Add>::Output as Sub<U1>>::Output: ArrayLength,
         <λ as Sub<U1>>::Output: ArrayLength,
+        I: IntoIterator<Item = F>,
+        J: IntoIterator<Item = F>,
+        I::IntoIter: ExactSizeIterator,
+        J::IntoIter: ExactSizeIterator,
     {
-        debug_assert_eq!(self.u.len() % λ::USIZE, 0); // We should pad with zeroes eventually
+        let mut u = u.into_iter();
+        let mut v = v.into_iter();
+
+        debug_assert_eq!(u.len() % λ::USIZE, 0); // We should pad with zeroes eventually
 
-        let s = self.u.len() / λ::USIZE;
+        let s = u.len() / λ::USIZE;
 
         assert!(
             s > 1,
             "When the output is this small, you should call `compute_final_proof`"
         );
 
-        let mut next_proof_generator = ProofGenerator {
-            u: Vec::<F>::with_capacity(s),
-            v: Vec::<F>::with_capacity(s),
-        };
-
         let denominator = CanonicalLagrangeDenominator::<F, λ>::new();
-        let lagrange_table_r = LagrangeTable::<F, λ, U1>::new(&denominator, &r);
         let lagrange_table = LagrangeTable::<F, λ, <λ as Sub<U1>>::Output>::from(denominator);
```

Comment:
the denominator and table only depend on lambda, so you don't need to recompute them for every proof part, you could add a reference to lagrange table to the input

Reply:
You're right. I was planning to do that but forgot. Thanks for the reminder! I'll add it.